File: src/config/presets.ts
/**
 * 供应商目录（presets）加载层。
 *
 * 加载优先级（评审⑤）：
 *   1) 用户目录 ~/.cc-mini/presets.json（可选，用户加供应商不必动 node_modules）
 *   2) 包内随发布的 presets.json（dist 的上级目录）
 *   3) 内置常量 BUILTIN_PRESETS（最后兜底，等价于现版 $BuiltinPresetsJson）
 *
 * 任一步解析失败/格式不对就跌落到下一步，绝不抛错中断启动。
 */
import { existsSync, readFileSync } from 'node:fs';
import { join } from 'node:path';
import { fileURLToPath } from 'node:url';

import { resolveStorePaths } from './store.js';
import type { Preset, PresetModels, PresetUrl } from './types.js';

export type { Preset } from './types.js';

/** 内置兜底目录（镜像仓库 presets.json）。第三方供应商**绝不**预置 `[1m]`（见 plan §3.1.1）。 */
export const BUILTIN_PRESETS: Preset[] = [
  {
    name: 'DeepSeek',
    auth: 'AUTH_TOKEN',
    effort: 'max',
    urls: [{ label: 'Anthropic 兼容', url: 'https://api.deepseek.com/anthropic' }],
    models: { opus: 'deepseek-v4-pro', sonnet: 'deepseek-v4-pro', haiku: 'deepseek-v4-flash' },
  },
  {
    name: '智谱GLM',
    auth: 'AUTH_TOKEN',
    urls: [{ label: 'Anthropic 兼容', url: 'https://open.bigmodel.cn/api/anthropic' }],
    models: { opus: 'GLM-4.7', sonnet: 'GLM-4.7', haiku: 'glm-4.5-air' },
  },
  {
    name: '小米MiMo',
    auth: 'AUTH_TOKEN',
    urls: [
      { label: '按量付费API', url: 'https://api.xiaomimimo.com/anthropic' },
      { label: 'TokenPlan', url: 'https://token-plan-cn.xiaomimimo.com/anthropic' },
    ],
    models: { opus: 'mimo-v2.5-pro', sonnet: 'mimo-v2.5-pro', haiku: 'mimo-v2.5-pro' },
  },
  {
    name: '官方Anthropic',
    auth: 'API_KEY',
    urls: [{ label: '(留空，用登录态)', url: '' }],
    models: {},
  },
];

const asString = (v: unknown): string => (typeof v === 'string' ? v : '');

function normalizeUrl(raw: unknown): PresetUrl {
  const u = (raw && typeof raw === 'object' ? raw : {}) as Record<string, unknown>;
  return { label: asString(u.label), url: asString(u.url) };
}

function normalizeModels(raw: unknown): PresetModels {
  const m = (raw && typeof raw === 'object' ? raw : {}) as Record<string, unknown>;
  const models: PresetModels = {};
  if (typeof m.opus === 'string') models.opus = m.opus;
  if (typeof m.sonnet === 'string') models.sonnet = m.sonnet;
  if (typeof m.haiku === 'string') models.haiku = m.haiku;
  return models;
}

function normalizePreset(raw: unknown): Preset | undefined {
  const p = (raw && typeof raw === 'object' ? raw : {}) as Record<string, unknown>;
  const name = asString(p.name).trim();
  if (!name) return undefined; // 无名条目直接丢弃
  const auth: Preset['auth'] = p.auth === 'API_KEY' ? 'API_KEY' : 'AUTH_TOKEN';
  const urls = Array.isArray(p.urls) ? p.urls.map(normalizeUrl) : [];
  const preset: Preset = { name, auth, urls, models: normalizeModels(p.models) };
  if (typeof p.effort === 'string' && p.effort.trim()) preset.effort = p.effort.trim();
  return preset;
}

/** 把任意解析结果规整为 Preset[]；非数组或全空则返回 undefined（让调用方跌落兜底）。 */
function normalizePresets(raw: unknown): Preset[] | undefined {
  if (!Array.isArray(raw)) return undefined;
  const list = raw.map(normalizePreset).filter((p): p is Preset => p !== undefined);
  return list.length > 0 ? list : undefined;
}

/** 尝试读并解析一个 presets.json；任何问题都安静返回 undefined。 */
function tryLoadFile(file: string): Preset[] | undefined {
  try {
    if (!existsSync(file)) return undefined;
    return normalizePresets(JSON.parse(readFileSync(file, 'utf-8')));
  } catch {
    return undefined;
  }
}

/** 定位包内随发布的 presets.json（dist 的上级 = 包根；dev 下 = 仓库根）。 */
function packagePresetsFile(): string {
  // 本模块编译后位于 dist/config/presets.js → ../../presets.json = 包根。
  return fileURLToPath(new URL('../../presets.json', import.meta.url));
}

/**
 * 加载供应商目录。`storeDir` 来自 --store-dir（测试用），决定用户覆盖文件的位置。
 * 优先级见文件头注释。
 */
export function loadPresets(storeDir?: string): Preset[] {
  const userFile = join(resolveStorePaths(storeDir).dir, 'presets.json');
  return tryLoadFile(userFile) ?? tryLoadFile(packagePresetsFile()) ?? BUILTIN_PRESETS;
}

File: internal/presets/presets.go
// Package presets 是供应商目录（presets）加载层：用户覆盖文件 > 二进制旁路文件 > 内置兜底。
//
// 术语：一个保存的条目叫「配置」(profile，见 internal/config)，这里的目录条目叫「供应商」(provider)。
// 与 internal/config 解耦：presets 可以依赖 config（取存储目录），但 config 不依赖 presets。
//
// 实现说明：计划 §4.2 提到 go:embed，但 Go embed 无法引用包外的 ../../presets.json（npm 发布用的根文件）。
// 为避免在 internal/presets 再放一份同名副本造成漂移，这里改用与 TS 一致的「字面量 BuiltinPresets +
// 对拍测试（presets_test.go 断言它等于根 presets.json）」方案，根目录仍是唯一可编辑源。
package presets

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"

	"github.com/becomeless/cc-x/internal/config"
)

// 认证字段取值。
const (
	AuthToken  = "AUTH_TOKEN"
	AuthAPIKey = "API_KEY"
)

// URL 是某供应商的一个 API 地址（可多个，多个时让用户选）。
type URL struct {
	Label string `json:"label"`
	URL   string `json:"url"`
}

// Models 是三档模型映射（可部分为空）。
type Models struct {
	Opus   string `json:"opus,omitempty"`
	Sonnet string `json:"sonnet,omitempty"`
	Haiku  string `json:"haiku,omitempty"`
}

// Preset 是一个「供应商」目录条目。
type Preset struct {
	Name   string `json:"name"`
	Auth   string `json:"auth"` // AUTH_TOKEN | API_KEY
	URLs   []URL  `json:"urls"`
	Models Models `json:"models"`
	Effort string `json:"effort,omitempty"`
}

// BuiltinPresets 是内置兜底目录，镜像仓库根 presets.json（由 presets_test.go 对拍保证不漂）。
// 第三方供应商绝不预置 `[1m]`（见 plan §3.1.1）。
var BuiltinPresets = []Preset{
	{
		Name:   "DeepSeek",
		Auth:   AuthToken,
		Effort: "max",
		URLs:   []URL{{Label: "Anthropic 兼容", URL: "https://api.deepseek.com/anthropic"}},
		Models: Models{Opus: "deepseek-v4-pro", Sonnet: "deepseek-v4-pro", Haiku: "deepseek-v4-flash"},
	},
	{
		Name:   "智谱GLM",
		Auth:   AuthToken,
		URLs:   []URL{{Label: "Anthropic 兼容", URL: "https://open.bigmodel.cn/api/anthropic"}},
		Models: Models{Opus: "GLM-4.7", Sonnet: "GLM-4.7", Haiku: "glm-4.5-air"},
	},
	{
		Name: "小米MiMo",
		Auth: AuthToken,
		URLs: []URL{
			{Label: "按量付费API", URL: "https://api.xiaomimimo.com/anthropic"},
			{Label: "TokenPlan", URL: "https://token-plan-cn.xiaomimimo.com/anthropic"},
		},
		Models: Models{Opus: "mimo-v2.5-pro", Sonnet: "mimo-v2.5-pro", Haiku: "mimo-v2.5-pro"},
	},
	{
		Name:   "官方Anthropic",
		Auth:   AuthAPIKey,
		URLs:   []URL{{Label: "(留空，用登录态)", URL: ""}},
		Models: Models{},
	},
}

func asString(v any) string {
	if s, ok := v.(string); ok {
		return s
	}
	return ""
}

func normalizeURL(raw any) URL {
	m, _ := raw.(map[string]any)
	return URL{Label: asString(m["label"]), URL: asString(m["url"])}
}

func normalizeModels(raw any) Models {
	m, _ := raw.(map[string]any)
	return Models{Opus: asString(m["opus"]), Sonnet: asString(m["sonnet"]), Haiku: asString(m["haiku"])}
}

// normalizePreset 宽松规整一条；无名条目返回 ok=false 由调用方丢弃。
func normalizePreset(raw any) (Preset, bool) {
	m, _ := raw.(map[string]any)
	name := strings.TrimSpace(asString(m["name"]))
	if name == "" {
		return Preset{}, false
	}
	auth := AuthToken
	if asString(m["auth"]) == AuthAPIKey {
		auth = AuthAPIKey
	}
	urls := []URL{}
	if arr, ok := m["urls"].([]any); ok {
		for _, u := range arr {
			urls = append(urls, normalizeURL(u))
		}
	}
	p := Preset{Name: name, Auth: auth, URLs: urls, Models: normalizeModels(m["models"])}
	if e := strings.TrimSpace(asString(m["effort"])); e != "" {
		p.Effort = e
	}
	return p, true
}

// normalizePresets 把任意解析结果规整为 []Preset；非数组或全空返回 nil（让调用方跌落兜底）。
func normalizePresets(raw any) []Preset {
	arr, ok := raw.([]any)
	if !ok {
		return nil
	}
	out := make([]Preset, 0, len(arr))
	for _, item := range arr {
		if p, ok := normalizePreset(item); ok {
			out = append(out, p)
		}
	}
	if len(out) == 0 {
		return nil
	}
	return out
}

// tryLoadFile 尝试读并解析一个 presets.json；任何问题都安静返回 nil。
func tryLoadFile(file string) []Preset {
	data, err := os.ReadFile(file)
	if err != nil {
		return nil
	}
	var raw any
	if json.Unmarshal(data, &raw) != nil {
		return nil
	}
	return normalizePresets(raw)
}

// Load 加载供应商目录。优先级：用户 <storeDir>/presets.json > 二进制旁路 presets.json > 内置 BuiltinPresets。
// 任一步缺失/损坏都安静跌落下一步，绝不中断启动。
func Load(storeDir string) []Preset {
	userFile := filepath.Join(config.ResolveStorePaths(storeDir).Dir, "presets.json")
	if p := tryLoadFile(userFile); p != nil {
		return p
	}
	if exe, err := os.Executable(); err == nil {
		if p := tryLoadFile(filepath.Join(filepath.Dir(exe), "presets.json")); p != nil {
			return p
		}
	}
	return BuiltinPresets
}
